File: 206 - Reverse Linked List/solution.java


class Solution {
    // Definition for singly-linked list.
    public class ListNode {
        int val;
        ListNode next;
        ListNode() {}
        ListNode(int val) { this.val = val; }
        ListNode(int val, ListNode next) { this.val = val; this.next = next; }
    }

    public ListNode reverseList(ListNode head) {
        ListNode prev = null;
        ListNode curr = head;

        while(curr != null){
            ListNode next = curr.next;
            curr.next = prev;
            
            prev = curr;
            curr = next;
        }

        // for recurssive
        // return recurssive(head);
        return prev;
    }

    // Using recurssion
    public ListNode recurssive(ListNode head){
        if(head==null || head.next == null){
            return head;
        }

        ListNode newHead = recurssive(head.next);

        ListNode front = head.next;
        front.next = head;
        head.next = null;
        
        return newHead;
    }
}

File: 206 - Reverse Linked List/solution.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var reverseList = function(head) {
    let prev = null;
    let curr = head;

    while(curr) {
        // setting the next element in next node
        let next = curr.next;
        // setting the next of curr to previous node
        curr.next = prev;
        // setting the prev node as the current
        prev = curr;
        // setting the current as next
        curr = next;
    }

    // Lastly the prev would point to the last node
    return prev;
};

// Using recurssion
var recurssive = function(head){
    const reverse = (head) => {
        if(head == null || head.next == null){
            return head;
        }

        let newHead = reverse(head.next);
        let front = head.next;
        front.next = head;
        head.next = null;
        return newHead;

    }

    return reverse(head);
}


